File: Practical-5/binomial_heap.cpp
#include <bits/stdc++.h>
using namespace std;
template <typename T>
class BinomialHeap
{
public:
    struct Node
    {
        T key;
        int degree;
        Node *parent, *child, *sibling;
        Node(T k) : key(k), degree(0), parent(nullptr), child(nullptr), sibling(nullptr) {}
    };

private:
    Node *head;
    void linkTrees(Node *y, Node *z)
    {
        y->parent = z;
        y->sibling = z->child;
        z->child = y;
        z->degree += 1;
    }
    static Node *mergeRootLists(Node *h1, Node *h2)
    {
        if (!h1)
            return h2;
        if (!h2)
            return h1;
        Node *head = nullptr;
        Node *tail = nullptr;
        Node *a = h1;
        Node *b = h2;
        while (a && b)
        {
            Node *next;
            if (a->degree <= b->degree)
            {
                next = a;
                a = a->sibling;
            }
            else
            {
                next = b;
                b = b->sibling;
            }
            if (!head)
                head = tail = next;
            else
            {
                tail->sibling = next;
                tail = next;
            }
        }
        Node *remain = (a ? a : b);
        if (tail)
            tail->sibling = remain;
        else
            head = remain;
        return head;
    }
    static Node *unionHeaps(Node *h1, Node *h2)
    {
        Node *newHead = mergeRootLists(h1, h2);
        if (!newHead)
            return nullptr;
        Node *prev = nullptr;
        Node *curr = newHead;
        Node *next = curr->sibling;
        while (next)
        {
            if (curr->degree != next->degree ||
                (next->sibling && next->sibling->degree == curr->degree))
            {
                prev = curr;
                curr = next;
            }
            else
            {
                if (curr->key <= next->key)
                {
                    curr->sibling = next->sibling;
                    linkTrees(next, curr);
                }
                else
                {
                    if (!prev)
                        newHead = next;
                    else
                        prev->sibling = next;
                    linkTrees(curr, next);
                    curr = next;
                }
            }
            next = curr->sibling;
        }
        return newHead;
    }
    static Node *reverseList(Node *node)
    {
        Node *prev = nullptr;
        while (node)
        {
            Node *nxt = node->sibling;
            node->sibling = prev;
            node->parent = nullptr;
            prev = node;
            node = nxt;
        }
        return prev;
    }
    Node *findNode(Node *root, const T &key)
    {
        if (!root)
            return nullptr;
        if (root->key == key)
            return root;
        Node *res = findNode(root->child, key);
        if (res)
            return res;
        return findNode(root->sibling, key);
    }

public:
    BinomialHeap() : head(nullptr) {}
    ~BinomialHeap()
    {
        function<void(Node *)> deleteAll = [&](Node *node)
        {
            if (!node)
                return;
            deleteAll(node->child);
            deleteAll(node->sibling);
            delete node;
        };
        deleteAll(head);
    }
    bool isEmpty() const { return head == nullptr; }
    void insert(const T &key)
    {
        Node *single = new Node(key);
        head = unionHeaps(head, single);
    }
    void merge(BinomialHeap<T> &other)
    {
        head = unionHeaps(head, other.head);
        other.head = nullptr;
    }
    T getMin() const
    {
        if (!head)
            throw runtime_error("Heap is empty");
        Node *y = nullptr;
        Node *x = head;
        T minVal;
        bool first = true;
        while (x)
        {
            if (first || x->key < minVal)
            {
                minVal = x->key;
                y = x;
                first = false;
            }
            x = x->sibling;
        }
        return minVal;
    }
    T extractMin()
    {
        if (!head)
            throw runtime_error("Heap is empty");
        Node *prevMin = nullptr;
        Node *minNode = head;
        Node *prev = nullptr;
        Node *cur = head;
        T minKey = cur->key;
        while (cur)
        {
            if (cur->key < minKey)
            {
                minKey = cur->key;
                prevMin = prev;
                minNode = cur;
            }
            prev = cur;
            cur = cur->sibling;
        }
        if (!prevMin)
            head = minNode->sibling;
        else
            prevMin->sibling = minNode->sibling;
        Node *childHead = reverseList(minNode->child);
        T retKey = minNode->key;
        delete minNode;
        head = unionHeaps(head, childHead);
        return retKey;
    }
    Node *find(const T &key)
    {
        return findNode(head, key);
    }
    void decreaseKey(Node *node, const T &newKey)
    {
        if (!node)
            throw runtime_error("Null node passed to decreaseKey");
        if (newKey > node->key)
            throw runtime_error("New key is greater than current key in decreaseKey");
        node->key = newKey;
        Node *y = node;
        Node *z = y->parent;
        while (z && y->key < z->key)
        {
            swap(y->key, z->key);
            y = z;
            z = y->parent;
        }
    }
    bool deleteByValue(const T &key)
    {
        Node *node = find(key);
        if (!node)
            return false;
        T sentinel = std::numeric_limits<T>::lowest();
        decreaseKey(node, sentinel);
        extractMin();
        return true;
    }
    void print() const
    {
        cout << "Binomial Heap root list (degree : keys at root):\n";
        Node *cur = head;
        while (cur)
        {
            cout << "degree " << cur->degree << " -> " << cur->key << " ; ";
            cur = cur->sibling;
        }
        cout << "\n(Use detailedPrint() for tree-by-tree view)\n";
    }
    void detailedPrint() const
    {
        function<void(const Node *, int)> dfs = [&](const Node *n, int indent)
        {
            if (!n)
                return;
            for (int i = 0; i < indent; i++)
                cout << "  ";
            cout << n->key << " (deg=" << n->degree << ")\n";
            dfs(n->child, indent + 1);
            dfs(n->sibling, indent);
        };
        cout << "Detailed Binomial Heap:\n";
        Node *cur = head;
        while (cur)
        {
            cout << "Tree root: " << cur->key << " (degree " << cur->degree << ")\n";
            dfs(cur->child, 1);
            cur = cur->sibling;
        }
    }
};
int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    BinomialHeap<int> bh;
    bh.insert(10);
    bh.insert(3);
    bh.insert(7);
    bh.insert(1);
    bh.insert(5);
    cout << "Min: " << bh.getMin() << "\n";
    bh.print();
    cout << "Extracted min: " << bh.extractMin() << "\n";
    cout << "New min: " << bh.getMin() << "\n";
    auto node10 = bh.find(10);
    if (node10)
    {
        bh.decreaseKey(node10, 2);
        cout << "After decreasing 10->2, min: " << bh.getMin() << "\n";
    }
    bool deleted = bh.deleteByValue(7);
    cout << "Delete 7: " << (deleted ? "success" : "not found") << "\n";
    bh.detailedPrint();
    BinomialHeap<int> other;
    other.insert(4);
    other.insert(8);
    bh.merge(other);
    cout << "After merge, min: " << bh.getMin() << "\n";
    bh.detailedPrint();
    cout << "Extracting all: ";
    while (!bh.isEmpty())
    {
        cout << bh.extractMin() << " ";
    }
    cout << "\n";
    return 0;
}

File: data_strucs/binomial_heap.java
import java.util.*;

public class binomial_heap {
// Node structure for the Binomial Heap
struct Node {
    int key;
    int degree;
    Node *parent, *child, *sibling;

    Node(int k) : key(k), degree(0), parent(null), child(null), sibling(null) {}
};

class BinomialHeap {
public:
    Node* head; // Points to the head of the root list

    // Link two binomial trees of the same degree
    void linkTrees(Node* y, Node* z) {
        y.parent = z;
        y.sibling = z.child;
        z.child = y;
        z.degree++;
    }

    // Merge the root lists of two heaps into a single sorted list by degree
    Node* mergeHeaps(Node* h1, Node* h2) {
        if (!h1) return h2;
        if (!h2) return h1;

        Node* newHead = null;
        Node* tail = null;

        Node* curr1 = h1;
        Node* curr2 = h2;

        if (curr1.degree <= curr2.degree) {
            newHead = curr1;
            curr1 = curr1.sibling;
        } else {
            newHead = curr2;
            curr2 = curr2.sibling;
        }
        tail = newHead;

        while (curr1 && curr2) {
            if (curr1.degree <= curr2.degree) {
                tail.sibling = curr1;
                curr1 = curr1.sibling;
            } else {
                tail.sibling = curr2;
                curr2 = curr2.sibling;
            }
            tail = tail.sibling;
        }

        // Append the remaining list
        if (curr1) tail.sibling = curr1;
        else tail.sibling = curr2;
        
        return newHead;
    }

public:
    // Constructor
    BinomialHeap() : head(null) {}

    // Function to check if the heap is empty
    boolean isEmpty() {
        return head == null;
    }

    // Find the node with the minimum key in the root list
    Node* findMinimum() {
        if (!head) return null;

        Node* minNode = head;
        Node* current = head.sibling;
        while (current) {
            if (current.key < minNode.key) {
                minNode = current;
            }
            current = current.sibling;
        }
        return minNode;
    }

    // Unite two binomial heaps
    Node* unionHeaps(Node* h1, Node* h2) {
        Node* newHead = mergeHeaps(h1, h2);
        if (!newHead) return null;

        Node* prev = null;
        Node* x = newHead;
        Node* next = x.sibling;

        while (next) {
            // Case 1 & 2: Degrees are different or three consecutive trees have same degree
            if ((x.degree != next.degree) || (next.sibling && next.sibling->degree == x.degree)) {
                prev = x;
                x = next;
            } else {
                // Case 3 & 4: Two consecutive trees have same degree
                if (x.key <= next.key) {
                    x.sibling = next.sibling;
                    linkTrees(next, x);
                } else {
                    if (!prev) newHead = next;
                    else prev.sibling = next;
                    linkTrees(x, next);
                    x = next;
                }
            }
            next = x.sibling;
        }
        return newHead;
    }

    // Insert a key into the heap
    void insert(int key) {
        Node* newNode = new Node(key);
        BinomialHeap tempHeap;
        tempHeap.head = newNode;
        head = unionHeaps(this.head, tempHeap.head);
    }

    // Extract the node with the minimum key
    int extractMin() {
        if (!head) return -1; // Or throw exception

        Node* minNode = findMinimum();
        int minKey = minNode.key;

        // Remove minNode from the root list
        Node* prev = null;
        Node* current = head;
        while (current != minNode) {
            prev = current;
            current = current.sibling;
        }
        if (!prev) head = minNode.sibling;
        else prev.sibling = minNode.sibling;

        // Create a new heap from the children of the minNode
        Node* childHead = null;
        Node* child = minNode.child;
        while (child) {
            Node* nextChild = child.sibling;
            child.sibling = childHead;
            child.parent = null;
            childHead = child;
            child = nextChild;
        }

        // Union the original heap with the new heap
        BinomialHeap tempHeap;
        tempHeap.head = childHead;
        head = unionHeaps(this.head, tempHeap.head);

        delete minNode;
        return minKey;
    }
    
    // Decrease the key of a given node
    void decreaseKey(Node* node, int newKey) {
        if (!node || newKey > node.key) return;

        node.key = newKey;
        Node* y = node;
        Node* z = y.parent;

        // Bubble the key up to maintain heap property
        while (z && y.key < z.key) {
            swap(y.key, z.key);
            // If we need to track nodes, we'd swap more than just the key
            y = z;
            z = y.parent;
        }
    }

    // Delete a given node from the heap
    void deleteNode(Node* node) {
        if (!node) return;
        
        // Decrease key to negative infinity and extract it
        decreaseKey(node, std::numeric_limits<int>::min());
        extractMin();
    }
    
    // Simple print function to show the keys in the root list
    void printRootList() {
        if (!head) {
            std::System.out.print( "Heap is empty." );
            return;
        }
        Node* current = head;
        std::System.out.print( "Root list: ";
        while(current) {
            std::System.out.print( current.key << "(B" << current.degree << ") ";
            current = current.sibling;
        }
        std::System.out.print( "\n";
    }
};

public static void main(String[] args) {
    BinomialHeap heap;

    // --- Insert Demo ---
    std::System.out.print( "--- Insertion Demo ---" );
    heap.insert(10);
    heap.insert(20);
    heap.insert(5); // New minimum
    heap.insert(30);
    heap.insert(3); // New minimum
    heap.printRootList();

    // --- Find Minimum Demo ---
    std::System.out.print( "\n--- Find Minimum ---" );
    Node* minNode = heap.findMinimum();
    if (minNode) {
        std::System.out.print( "Minimum key is: " << minNode.key );
    }

    // --- Extract Minimum Demo ---
    std::System.out.print( "\n--- Extract Minimum ---" );
    std::System.out.print( "Extracted min: " << heap.extractMin() );
    heap.printRootList();
    minNode = heap.findMinimum();
    if (minNode) {
        std::System.out.print( "New minimum key is: " << minNode.key );
    }
    
    // --- Union Demo ---
    std::System.out.print( "\n--- Union Demo ---" );
    BinomialHeap heap2;
    heap2.insert(7);
    heap2.insert(18);
    heap2.insert(22);
    std::System.out.print( "Heap 1: "; heap.printRootList();
    std::System.out.print( "Heap 2: "; heap2.printRootList();
    
    BinomialHeap combinedHeap;
    // HACK: For simplicity, we directly access heads. A proper API would be cleaner.
    combinedHeap.head = combinedHeap.unionHeaps(heap.head, heap2.head);
    std::System.out.print( "Combined Heap: "; combinedHeap.printRootList();
    minNode = combinedHeap.findMinimum();
    if (minNode) {
        std::System.out.print( "Minimum key of combined heap is: " << minNode.key );
    }

    // --- Decrease Key and Delete Demo ---
    std::System.out.print( "\n--- Decrease Key & Delete ---" );
    // In a real application, you would need a way to get a pointer to a node.
    // For this demo, we'll find a node to work with. Let's find node with key 30.
    Node* nodeToModify = combinedHeap.head; // Should be the B3 tree root (30)
    while (nodeToModify && nodeToModify.key != 30) {
        nodeToModify = nodeToModify.sibling;
    }

    if (nodeToModify) {
        std::System.out.print( "Decreasing key of node 30 to 1." );
        combinedHeap.decreaseKey(nodeToModify, 1);
        minNode = combinedHeap.findMinimum();
        std::System.out.print( "New minimum key is: " << minNode.key );
        combinedHeap.printRootList();

        std::System.out.print( "\nDeleting node with key 22." );
        Node* nodeToDelete = combinedHeap.head; // Should be B0 tree root (22)
         while (nodeToDelete && nodeToDelete.key != 22) {
            nodeToDelete = nodeToDelete.sibling;
        }
        if(nodeToDelete) {
            combinedHeap.deleteNode(nodeToDelete);
            std::System.out.print( "Heap after deleting 22: "; combinedHeap.printRootList();
        } else {
             std::System.out.print( "Node 22 not found in root list." );
        }

    } else {
        std::System.out.print( "Node with key 30 not found in root list for demo." );
    }

    return 0;
}
}